File: packages/utils/src/run-differential-fuzz.py
# -*- coding: utf-8 -*-
"""
ZTAN Cross-Runtime Differential Fuzzing Bridge (Python)
Accepts JSON blocks or raw strings via stdin, runs duplicate key detection and 
JCS canonicalization, and returns deterministic outcome markers.
"""

import sys
import json
import hashlib
import unicodedata
import re
import math
import io

# Force standard streams to use strict UTF-8 encoding across all OS architectures
sys.stdin = io.TextIOWrapper(sys.stdin.buffer, encoding='utf-8')
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

MAX_DEPTH = 64
MAX_BYTES = 1_000_000
MAX_KEYS = 10_000
MAX_STRING = 65_536

def has_lone_surrogate(val):
    if not isinstance(val, str):
        return False
    return any(55296 <= ord(c) <= 57343 for c in val)

def validate_duplicate_keys(json_str):
    if len(json_str) > MAX_BYTES:
        raise ValueError(f'[JCS] JSON payload of {len(json_str)} bytes exceeds MAX_BYTES buffer of {MAX_BYTES}.')

    stack = [set()]
    i = 0
    while i < len(json_str):
        if len(stack) > MAX_DEPTH:
            raise ValueError(f'[JCS] Parser nesting depth limit of {MAX_DEPTH} exceeded.')

        char = json_str[i]
        if char == '{':
            stack.append(set())
            i += 1
        elif char == '}':
            if len(stack) > 1:
                stack.pop()
            i += 1
        elif char == '"':
            key_val = ''
            i += 1
            while i < len(json_str) and json_str[i] != '"':
                if json_str[i] == '\\':
                    key_val += json_str[i] + (json_str[i + 1] if i + 1 < len(json_str) else '')
                    i += 2
                else:
                    key_val += json_str[i]
                    i += 1

                if len(key_val) > MAX_STRING:
                    raise ValueError(f'[JCS] Parsed string token exceeds safety limit of {MAX_STRING} characters.')
            i += 1
            if has_lone_surrogate(key_val):
                raise ValueError('[JCS] Lone surrogates are strictly forbidden to prevent UTF-8 encoding anomalies.')

            next_idx = i
            while next_idx < len(json_str) and json_str[next_idx].isspace():
                next_idx += 1

            if next_idx < len(json_str) and json_str[next_idx] == ':':
                current_set = stack[-1]
                if len(current_set) >= MAX_KEYS:
                    raise ValueError(f'[JCS] Lexical keys count in scope exceeds safe limit of {MAX_KEYS}.')
                if key_val in current_set:
                    raise ValueError(f'[JCS] Duplicate JSON key detected: "{key_val}". Block is malformed.')
                current_set.add(key_val)
                i = next_idx + 1
            else:
                i = next_idx
        elif char.isalpha():
            word = ''
            while i < len(json_str) and json_str[i].isalpha():
                word += json_str[i]
                i += 1
            if word not in ('true', 'false', 'null', 'e', 'E'):
                raise ValueError(f'[JCS] Invalid unquoted literal or identifier detected: "{word}".')
        else:
            i += 1

def canonicalize_jcs(value, seen_objects=None, depth=0):
    if seen_objects is None:
        seen_objects = set()

    if depth > MAX_DEPTH:
        raise ValueError(f'[JCS] Maximum nesting depth of {MAX_DEPTH} exceeded.')

    if value is None:
        return 'null'

    if isinstance(value, str):
        if len(value) > MAX_STRING:
            raise ValueError(f'[JCS] Maximum string length of {MAX_STRING} exceeded.')
        if has_lone_surrogate(value):
            raise ValueError('[JCS] Lone surrogates are strictly forbidden to prevent UTF-8 encoding anomalies.')
        normalized = unicodedata.normalize('NFC', value)
        # Match Javascript's JSON.stringify escaping
        escaped = json.dumps(normalized, ensure_ascii=False)
        if len(escaped) > MAX_STRING * 2:
            raise ValueError('[JCS] Escaped string value exceeds maximum serialization safety limits.')
        return escaped

    # Check for boolean first since isinstance(True, int) is True in Python
    if isinstance(value, bool):
        return 'true' if value else 'false'

    if isinstance(value, (int, float)):
        if not math.isfinite(value):
            raise ValueError('[JCS] Non-finite number values (Infinity, NaN) are strictly forbidden.')
        
        # Format matching JCS specifications
        val_str = str(value)
        if val_str.endswith('.0'):
            val_str = val_str[:-2]
        # Match scientific notation boundary format
        if 'e' in val_str or 'E' in val_str:
            # Let JSON serialize standard format
            val_str = json.dumps(value)
        return val_str

    if isinstance(value, list):
        items = [canonicalize_jcs(item, seen_objects, depth + 1) for item in value]
        result = '[' + ','.join(items) + ']'
        if len(result) > MAX_BYTES:
            raise ValueError(f'[JCS] Serialized array output exceeds maximum byte limit of {MAX_BYTES}.')
        return result

    if isinstance(value, dict):
        obj_id = id(value)
        if obj_id in seen_objects:
            raise ValueError('[JCS] Circular structure detected; serialization aborted.')
        seen_objects.add(obj_id)

        sorted_keys = sorted(value.keys())
        if len(sorted_keys) > MAX_KEYS:
            raise ValueError(f'[JCS] Maximum object keys limit of {MAX_KEYS} exceeded in a single scope.')

        parts = []
        for key in sorted_keys:
            if len(key) > MAX_STRING:
                raise ValueError(f'[JCS] Object key exceeds maximum length of {MAX_STRING}.')
            if has_lone_surrogate(key):
                raise ValueError('[JCS] Lone surrogates are strictly forbidden to prevent UTF-8 encoding anomalies.')
            canonical_key = json.dumps(unicodedata.normalize('NFC', key), ensure_ascii=False)
            canonical_val = canonicalize_jcs(value[key], seen_objects, depth + 1)
            parts.append(f'{canonical_key}:{canonical_val}')

        seen_objects.remove(obj_id)
        result = '{' + ','.join(parts) + '}'
        if len(result) > MAX_BYTES:
            raise ValueError(f'[JCS] Serialized object output exceeds maximum byte limit of {MAX_BYTES}.')
        return result

    raise ValueError(f'[JCS] Unsupported data type: {type(value)}')

def decode_placeholders(val):
    if val == 'INFINITY_PLACEHOLDER':
        return float('inf')
    if val == 'NEG_INFINITY_PLACEHOLDER':
        return float('-inf')
    if val == 'NAN_PLACEHOLDER':
        return float('nan')
    if isinstance(val, dict):
        return {k: decode_placeholders(v) for k, v in val.items()}
    if isinstance(val, list):
        return [decode_placeholders(v) for v in val]
    return val

def main():
    try:
        raw_input = sys.stdin.read()
        envelope = json.loads(raw_input)
        action = envelope.get('action')
        payload = decode_placeholders(envelope.get('payload'))

        if action == 'validate_duplicate_keys':
            validate_duplicate_keys(payload)
            print("OK")
            sys.exit(0)
        elif action == 'canonicalize_jcs':
            jcs_str = canonicalize_jcs(payload)
            print(f"OK:{jcs_str}")
            sys.exit(0)
        else:
            print(f"ERROR: Unknown action {action}")
            sys.exit(2)
    except Exception as e:
        print(f"ERROR:{str(e)}")
        sys.exit(1)

if __name__ == '__main__':
    main()

File: packages/utils/src/run-differential-fuzz.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath } from 'node:url';
import { spawnSync } from 'node:child_process';
import { canonicalizeJCS, validateDuplicateKeys } from './canonicalizer.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const PYTHON_FUZZER_BRIDGE = path.join(__dirname, 'run-differential-fuzz.py');

console.log('===========================================================');
console.log('    ZTAN Automated Cross-Runtime Differential Fuzzer        ');
console.log('===========================================================');

interface FuzzTestCase {
  id: string;
  category: string;
  action: 'validate_duplicate_keys' | 'canonicalize_jcs';
  payload: any; // Raw string for validate_duplicate_keys, JSON object/array/value for canonicalize_jcs
}

const testCases: FuzzTestCase[] = [];

// -----------------------------------------------------------------------------
// 1. Generate Fuzz Vectors targeting Edge-Case Vulnerability Spaces
// -----------------------------------------------------------------------------

// Category A: Duplicate Key Escaped Ambiguities & normalizations
const duplicateKeyEscapes = [
  { id: 'esc_hex_literal', raw: '{"\\u0061": 1, "a": 2}' },
  { id: 'esc_escaped_quote', raw: '{"a\\\\u0022": 1, "a\\"": 2}' },
  { id: 'esc_whitespace_tabs', raw: '{"\\tkey": 1, "key": 2}' },
  { id: 'esc_control_null', raw: '{"key\\u0000": 1, "key": 2}' },
  { id: 'esc_unicode_nfc_decomposed', raw: '{"a\\u0308": 1, "\\u00e4": 2}' },
  { id: 'esc_nested_dupes', raw: '{"nested": {"dup": 1, "dup": 2}}' },
  { id: 'esc_extreme_keys_limit', raw: '{' + Array.from({ length: 15 }, (_, i) => `"k${i}": ${i}`).join(',') + '}' },
];

for (const vector of duplicateKeyEscapes) {
  testCases.push({
    id: `dup_${vector.id}`,
    category: 'Duplicate Keys & Escapes',
    action: 'validate_duplicate_keys',
    payload: vector.raw
  });
}

// Category B: Float boundary limits
const floatBounds = [
  0.0,
  -0.0,
  1.0000000000000002, // JS float rounding border
  1e-20,
  1e+20,
  0.9850,
  45.0,
  -45.000,
  3.141592653589793,
];

for (let idx = 0; idx < floatBounds.length; idx++) {
  testCases.push({
    id: `float_bound_${idx}`,
    category: 'Float Limits & Normalization',
    action: 'canonicalize_jcs',
    payload: { payload: { rate: floatBounds[idx] } }
  });
}

// Category C: Non-finite numbers (should reject)
testCases.push({
  id: 'float_infinity',
  category: 'Non-Finite Floats',
  action: 'canonicalize_jcs',
  payload: { payload: { val: 'INFINITY_TRIGGER' } } // We'll inject Infinity dynamically
});

testCases.push({
  id: 'float_nan',
  category: 'Non-Finite Floats',
  action: 'canonicalize_jcs',
  payload: { payload: { val: 'NAN_TRIGGER' } } // We'll inject NaN dynamically
});

// Category D: Unicode glyph composition and surrogates
const unicodeGlyphs = [
  'a\u0308', // Decomposed ä
  '\u00e4', // Composed ä
  '💩', // Emoji surrogate pair
  '\uD83D', // Lone high surrogate
  '\uDE00', // Lone low surrogate
  '한국어', // Korean Hangul
  '👨‍👩‍👧‍👦', // Complex zero-width joiner family emoji
];

for (let idx = 0; idx < unicodeGlyphs.length; idx++) {
  testCases.push({
    id: `unicode_glyph_${idx}`,
    category: 'Unicode Composition & Surrogates',
    action: 'canonicalize_jcs',
    payload: { payload: { text: unicodeGlyphs[idx] } }
  });
}

// Category E: Resource Overflows (Depth, Strings, Bytes)
testCases.push({
  id: 'overflow_max_depth',
  category: 'Resource Boundaries',
  action: 'canonicalize_jcs',
  payload: 'DEPTH_OVERFLOW_TRIGGER'
});

testCases.push({
  id: 'overflow_max_string',
  category: 'Resource Boundaries',
  action: 'canonicalize_jcs',
  payload: { payload: { text: 'STRING_OVERFLOW_TRIGGER' } }
});

// Category F: Unquoted Identifiers & Non-Standard Literals (Duplicate Keys Scanner check)
const unquotedIdentifiers = [
  { id: 'infinity', raw: '{"rate": Infinity}' },
  { id: 'nan', raw: '{"rate": NaN}' },
  { id: 'undefined', raw: '{"rate": undefined}' },
  { id: 'malformed_word', raw: '{"rate": foo}' },
  { id: 'valid_exponent', raw: '{"rate": 1e5}' }, // Should PASS/CONVERGE perfectly
];

for (const vector of unquotedIdentifiers) {
  testCases.push({
    id: `unquoted_${vector.id}`,
    category: 'Unquoted Identifiers & Literals',
    action: 'validate_duplicate_keys',
    payload: vector.raw
  });
}

// -----------------------------------------------------------------------------
// 2. Differential Execution & Verification Orchestrator
// -----------------------------------------------------------------------------

let passedFuzzes = 0;
let totalFuzzes = 0;
let divergencesFound = 0;

interface DivergenceLog {
  id: string;
  category: string;
  payload: string;
  nodeResult: string;
  pyResult: string;
}
const divergenceLogs: DivergenceLog[] = [];

function runPythonBridge(action: string, payload: any): { status: 'OK' | 'ERROR'; value: string } {
  // Use a custom replacer to serialize Infinity and NaN safely as string placeholders to the Python subprocess
  const envelope = JSON.stringify({ action, payload }, (k, v) => {
    if (v === Infinity) return 'INFINITY_PLACEHOLDER';
    if (v === -Infinity) return 'NEG_INFINITY_PLACEHOLDER';
    if (typeof v === 'number' && Number.isNaN(v)) return 'NAN_PLACEHOLDER';
    return v;
  });

  const result = spawnSync('python', [PYTHON_FUZZER_BRIDGE], {
    input: envelope,
    encoding: 'utf8',
    env: { ...process.env, PYTHONIOENCODING: 'utf-8' }
  });

  if (result.error) {
    return { status: 'ERROR', value: `Python execution failed: ${result.error.message}` };
  }

  const output = (result.stdout || '').trim();
  if (output.startsWith('OK:')) {
    return { status: 'OK', value: output.slice(3) };
  } else if (output === 'OK') {
    return { status: 'OK', value: 'OK' };
  } else if (output.startsWith('ERROR:')) {
    return { status: 'ERROR', value: output.slice(6) };
  } else {
    return { status: 'ERROR', value: `Unexpected bridge output: ${output}\nStderr: ${result.stderr}` };
  }
}

for (const tc of testCases) {
  totalFuzzes++;
  let payload = tc.payload;

  // Perform dynamic mutation injections for triggers
  if (payload === 'DEPTH_OVERFLOW_TRIGGER') {
    let deepObj: any = {};
    let curr = deepObj;
    for (let d = 0; d < 70; d++) {
      curr.nest = {};
      curr = curr.nest;
    }
    payload = { payload: deepObj };
  } else if (tc.id === 'float_infinity') {
    payload.payload.val = Infinity;
  } else if (tc.id === 'float_nan') {
    payload.payload.val = NaN;
  } else if (payload?.payload?.text === 'STRING_OVERFLOW_TRIGGER') {
    payload.payload.text = 'x'.repeat(65537);
  }

  // 1. Run Node.js / TS validation
  let nodeStatus: 'OK' | 'ERROR' = 'OK';
  let nodeValue = '';

  try {
    if (tc.action === 'validate_duplicate_keys') {
      validateDuplicateKeys(payload);
      nodeValue = 'OK';
    } else {
      nodeValue = canonicalizeJCS(payload);
    }
  } catch (e: any) {
    nodeStatus = 'ERROR';
    nodeValue = e.message;
  }

  // 2. Run Python bridge validation
  const pyResult = runPythonBridge(tc.action, payload);

  // 3. Compare outcomes for perfect parity alignment
  const isParityAligned = (nodeStatus === pyResult.status) && 
    (nodeStatus === 'OK' ? nodeValue === pyResult.value : true); // If failed, error message wording can differ slightly, but action must align

  if (isParityAligned) {
    console.log(`[PASS] [${tc.category}] Vector ${tc.id} converged perfectly.`);
    passedFuzzes++;
  } else {
    console.error(`[DIVERGENCE] [${tc.category}] Vector ${tc.id} diverged!`);
    console.error(`  Node.js Status: ${nodeStatus} | Value: ${nodeValue}`);
    console.error(`  Python  Status: ${pyResult.status} | Value: ${pyResult.value}`);
    divergencesFound++;
    divergenceLogs.push({
      id: tc.id,
      category: tc.category,
      payload: JSON.stringify(payload),
      nodeResult: `Status: ${nodeStatus} | Value: ${nodeValue}`,
      pyResult: `Status: ${pyResult.status} | Value: ${pyResult.value}`
    });
  }
}

// -----------------------------------------------------------------------------
// 3. Differential Fuzzing Outcome Summary Matrix
// -----------------------------------------------------------------------------
console.log('\n===========================================================');
console.log('    ZTAN Differential Fuzzing Execution Matrix             ');
console.log('===========================================================');
console.log(`Vectors Executed:  ${totalFuzzes}`);
console.log(`Vectors Converged: ${passedFuzzes}`);
console.log(`Divergences Found: ${divergencesFound}`);
console.log('===========================================================');

if (divergencesFound > 0) {
  console.error('\n❌ CRITICAL: Parser Divergences Found between Node.js and Python!');
  for (const log of divergenceLogs) {
    console.error(`\nVector: ${log.id} (${log.category})`);
    console.error(`  Input Payload: ${log.payload}`);
    console.error(`  Node.js Result: ${log.nodeResult}`);
    console.error(`  Python Result:  ${log.pyResult}`);
  }
  process.exit(1);
} else {
  console.log('\n✅ SUCCESS: All fuzzing vectors converged perfectly across both runtimes.');
}
